### ISSUE TITLE
Application test harness

### ISSUE THREAD

**Issue by azriel91** [azriel91/autexousious#38]
In GitLab by @azriel91 on Jun 15, 2018, 11:43

In order to maintain quality at scale, there should be automated tests to ensure `System`s and their delegate components work as expected.

`System`s cannot easily be tested in unit test style tests, as well as any other components that work over `Storage`s, `Entity` objects, and `Component`s. Line code coverage is dropping because more systems and components are being added, and we don't have a good test harness to test them.

This issue should cover creation of a test harness that allows:

* Simple instantiation of an application
* Passing in `State`s
* Passing in assertion functions

**Out of scope:**

* Some way to drive the application through test code (e.g. can you put the application in a thread, and get it to read stdin?)

Ideally we can either re-use the same application instance, or run multiple applications in parallel without affecting each other.

---

Amethyst applications should also be run a sub thread due to [threading issues with mesa and GL software rendering](https://users.rust-lang.org/t/trouble-identifying-cause-of-segfault/18096), else a segfault tends to happen.

---

<details>
<summary>Proposed example usage</summary>

```rust
#[cfg(test)]
mod test {
    use amethyst_test_support::prelude::*;

    #[test]
    fn bundle_build_is_ok() {
        assert!(AmethystApplication::blank()
            .with_bundle(MyBundle)
            .run()
            .is_ok());
    }

    // May be needed for testing functions that need different systems registered
    #[test]
    fn load_multiple_bundles() {
        assert!(AmethystApplication::blank()
            .with_bundle(MyBundle0)
            .with_bundle(MyBundle1)
            .run()
            .is_ok());
    }

    #[test]
    fn bundle_registers_something() {
        let assertion_fn = |&mut world| {
            // Next line will panic if `Processor` wasn't added
            world.read_resource::<AssetStorage<Character>>();
        };
        assert!(AmethystApplication::blank()
            .with_bundle(MyBundle)
            .assertion(Box::new(assertion_fn))
            .run()
            .is_ok());
    }

    #[test]
    fn state_loads_something() {
        let assertion_fn = |&mut world| {
            // Next line will panic if `Processor` wasn't added
            world.read_resource::<AssetStorage<Character>>();
        };

        // Necessary if the State being tested is a loading state that returns `Trans::Switch`
        let assertion_state = AssertionState::new(Box::new(assertion_fn));

        // Contains animation, transform, ui, input bundles, but not the render bundle
        assert!(AmethystApplication::base()
            .with_state(LoadingState::new(Box::new(assertion_state)))
            .run()
            .is_ok());
    }

    #[test]
    fn function_creates_entity_with_component() {
        let assertion_fn = |&mut world| {
            let entity = MyEntitySpawner::spawn_for_player(
                world,
                // .. rest of args
            );

            // assert that the entity has the specified component
            assert!(
                world
                    .read_storage::<MyComponent>()
                    .contains(entity)
            );
        };

        assert!(AmethystApplication::base()
            .assertion(Box::new(assertion_fn))
            .run()
            .is_ok());
    }

    #[test]
    fn system_must_run_before_assertion() {
        let effect_fn = |world: &mut World| -> [MeshHandle; 2] {
            [
                SpriteMeshCreator::create_mesh(world, &sprites_definition()),
                SpriteMeshCreator::create_mesh_mirrored(world, &sprites_definition()),
            ]
        };
        let assertion_fn = |world: &mut World, mesh_handles: [MeshHandle; 2]| {
            let store = world.read_resource::<AssetStorage<Mesh>>();
            assert!(store.get(&mesh_handles[0]).is_some());
            assert!(store.get(&mesh_handles[1]).is_some());
        };

        assert!(AmethystApplication::base()
            .effect(Box::new(effect_fn))
            .assertion(Box::new(assertion_fn))
            .run()
            .is_ok());
    }
}
```

</details>

---

Useful when updating crates to use test harness:

```toml
[dev-dependencies]
amethyst_test_support = { path = "../amethyst_test_support" }

[features]
nightly = ["amethyst/nightly", "amethyst_test_support/nightly"]
```

```rust
#[cfg(test)]
extern crate amethyst_test_support;

use amethyst_test_support::prelude::*;

assert!(AmethystApplication::base().with_assertion().run().is_ok());
```
